tests/mutex.c

```diff
@@ -4,39 +4,82 @@
 #include <thread.h>
 #include <ktest.h>
 
-static mtx_t mtx = MTX_INITIALIZER(MTX_DEF);
-static volatile int32_t value;
+static mtx_t counter_mtx = MTX_INITIALIZER(MTX_DEF);
+static volatile int32_t counter_value;
 
-#define N 1000
-#define T 5
+#define COUNTER_N 1000
+#define COUNTER_T 5
 
-static thread_t *td[T];
+static thread_t *counter_td[COUNTER_T];
 
-static void mtx_test_main(void *arg) {
-  for (size_t i = 0; i < N; i++) {
-    mtx_lock(&mtx);
-    value++;
-    mtx_unlock(&mtx);
+static void counter_routine(void *arg) {
+  for (size_t i = 0; i < COUNTER_N; i++) {
+    mtx_lock(&counter_mtx);
+    int32_t v = counter_value;
+    thread_yield();
+    counter_value = v + 1;
+    mtx_unlock(&counter_mtx);
   }
 }
 
-static int mtx_test(void) {
-  value = 0;
+static int mtx_test_counter(void) {
+  counter_value = 0;
 
-  td[0] = thread_create("td1", mtx_test_main, NULL);
-  td[1] = thread_create("td2", mtx_test_main, NULL);
-  td[2] = thread_create("td3", mtx_test_main, NULL);
-  td[3] = thread_create("td4", mtx_test_main, NULL);
-  td[4] = thread_create("td5", mtx_test_main, NULL);
+  for (int i = 0; i < COUNTER_T; i++) {
+    char name[20];
+    snprintf(name, sizeof(name), "td%d", i);
+    counter_td[i] = thread_create(name, counter_routine, NULL);
+  }
+
+  for (int i = 0; i < COUNTER_T; i++)
+    sched_add(counter_td[i]);
+  for (int i = 0; i < COUNTER_T; i++)
+    thread_join(counter_td[i]);
+
+  assert(counter_value == COUNTER_N * COUNTER_T);
+
+  return KTEST_SUCCESS;
+}
+
+static mtx_t simple_mtx = MTX_INITIALIZER(MTX_DEF);
+static thread_t *simple_td0;
+static int simple_status = 0;
+
+static void simple_routine(void *arg) {
+  WITH_NO_PREEMPTION {
+    simple_status = 1;
+    mtx_lock(&simple_mtx);
+  }
+  WITH_NO_PREEMPTION {
+    simple_status = 2;
```

Comment:
Why do you need this `preempt_disable` and `simple_status = 2` here?

Reply:
Then I would delete
```
WITH_NO_PREEMPTION {
  simple_status = 2;
  mtx_unlock(&simple_mtx);
}
```
and leave only
```
mtx_unlock(&simple_mtx);
```
As you don't use value `2` of `simple_status` anywhere else in this test.

Reply:
I wish to leave it as it is for consistency. Actually value `2` has its meaning of not being `0`, `1` or `3`. Perhaps changing numbers to some enum:
```
enum status { INITIAL, LOCKING, UNLOCKING, DONE };
```
would make it more clear?

Reply:
That would be better but still -- in my opinion setting statuses that we don't even check is useless (and confusing for me as a code reader).